### ISSUE TITLE
LibJS: Assertion failed: index >= 0 in PropertyName constructor in ArrayPrototype::slice

### ISSUE THREAD

**Issue by Lubrsi** [SerenityOS/serenity#3932]
Found with Fuzzilli.

Code:
```js
function main() {
const v3 = [13.37];
const v4 = v3.slice(4294967297);
const v5 = [];
const v6 = {c:v5,d:v5,length:v5,toString:eval,valueOf:13.37};
}
main();
// Stderr:
// FuzzJs: /home/lukew/Desktop/serenity-project/serenity/Libraries/LibJS/Runtime/PropertyName.h:60: JS::PropertyName::PropertyName(i32): Assertion `index >= 0' failed.
```

Trace:
```
USERSPACE(24) ASSERTION FAILED: index >= 0
../Libraries/LibJS/Runtime/PropertyName.h:60
[js(24:24)]: Terminating js(24) due to signal 6
[FinalizerTask(3:3)]: CPU[0] NP(error) fault at invalid address V0x01e01d6c
[FinalizerTask(3:3)]: copy_from_user(0xc094a9ac, 0x01e01d6c, 4) failed at V0x01e01d6c
[FinalizerTask(3:3)]: 0xdeadc0de  Kernel::Processor::switch_context(Kernel::Thread*&, Kernel::Thread*&) +181
0xdeadc0de  Kernel::Scheduler::context_switch(Kernel::Thread*) +181
0xdeadc0de  Kernel::Scheduler::pick_next() +1965
0xdeadc0de  Kernel::Processor::check_invoke_scheduler() +125
0xdeadc0de  Kernel::Thread::die_if_needed() +157
0xdeadc0de  syscall_handler +1414
0xdeadc0de  syscall_asm_entry +49
0x080c3c78  raise +22
0x080480bd   +13
0x080bfa9b  tolower +0
0x0807f336  JS::PropertyName::PropertyName(int) +52
0x0809f324  JS::ArrayPrototype::slice(JS::VM&, JS::GlobalObject&) +322
0x0804bdd0  AK::Function<JS::Value (JS::VM&, JS::GlobalObject&)>::CallableWrapper<JS::Value (*)(JS::VM&, JS::GlobalObject&)>::call(JS::VM&, JS::GlobalObject&) const +22
0x080aa207  JS::NativeFunction::call() +59
0x0809622b  JS::VM::call_internal(JS::Function&, JS::Value, AK::Optional<JS::MarkedValueList>) +495
0x0805e870  JS::CallExpression::execute(JS::Interpreter&, JS::GlobalObject&) const +2340
0x0805e988  JS::VariableDeclaration::execute(JS::Interpreter&, JS::GlobalObject&) const +86
0x0806680d  JS::Interpreter::execute_statement(JS::GlobalObject&, JS::Statement const&, AK::Vector<JS::Argument, 8ul>, JS::ScopeType) +455
0x0808f354  JS::ScriptFunction::call() +1448
0x0809622b  JS::VM::call_internal(JS::Function&, JS::Value, AK::Optional<JS::MarkedValueList>) +495
0x0805e870  JS::CallExpression::execute(JS::Interpreter&, JS::GlobalObject&) const +2340
0x0805a05c  JS::ExpressionStatement::execute(JS::Interpreter&, JS::GlobalObject&) const +44
0x0806680d  JS::Interpreter::execute_statement(JS::GlobalObject&, JS::Statement const&, AK::Vector<JS::Argument, 8ul>, JS::ScopeType) +455
0x0805d520  JS::Program::execute(JS::Interpreter&, JS::GlobalObject&) const +64
0x08066c59  JS::Interpreter::run(JS::GlobalObject&, JS::Program const&) +407
0x0804b479  parse_and_run(JS::Interpreter&, AK::StringView const&) +467
0x08048df5  main +2310
0x08049a8e  _start +94
```

**Comment by linusg** [comment]
Simplified repro:

```
> [].slice((2 ** 32) / 2 - 1)
[  ]
> [].slice((2 ** 32) / 2)
js: ../Libraries/LibJS/Runtime/PropertyName.h:60: JS::PropertyName::PropertyName(i32): Assertion `index >= 0' failed.
```

---

This is integer overflow - max array index is supposed to be 2^32 - 2 (see [1]), and IndexedProperties supports up to `u32` but here we cast to `i32` which is incorrect:

https://github.com/SerenityOS/serenity/blob/7b863330dc72e5715b17726cdbae95c168015363/Libraries/LibJS/Runtime/ArrayPrototype.cpp#L386

This is a general issue in LibJS that needs to be addressed. We're not very spec compliant when it comes to numeric conversions and min/max values.

[1] https://tc39.es/ecma262/#sec-array-exotic-objects

> A String property name P is an array index if and only if ToString(ToUint32(P)) equals P and ToUint32(P) is not the same value as 𝔽(2^32 - 1).

**Comment by linusg** [comment]
This was fixed in #8036.
